marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/IOEndpointImpl.java

```diff
@@ -327,6 +344,25 @@ void assignEndpointState(InputStream[] output) {
             }
             return output;
         }
+
+        private void checkCallContextQueue() {
+            if(this.callContexts != null || !this.callContexts.isEmpty())
+                throw new InternalError("CallContext queue not empty");
+        }
+
+        class CallerThreadPoolExecutor extends ThreadPoolExecutor {
+
+            CallerThreadPoolExecutor(int threadCount, int queueSize) {
+
+                super(threadCount, threadCount, 0, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(queueSize),
+                        new CallerRunsPolicy());
+            }
+
+            private void submitTask(Callable<Boolean> callable) {
+                FutureTask futureTask = new FutureTask(callable);
+                super.execute((Runnable) callable);
```

Comment:
A Callable is not a Runnable, so this cast will throw a runtime error.

The FutureTask is a Runnable -- that's the reason for wrapping the callable in a FutureTask -- but the FutureTask is thrown away.

The task was suggesting to provide the submit method in the bulk caller instead of the ThreadPoolExecutor.
The submit() method in OutputEndpointImpl.BulkOutputCallerImpl looks correct. On seeing the BulkOutputCallerImpl.submit() method, I'm wondering whether a submit(Callable<Boolean> callable) method could be provided by BulkIOEndpointCallerImpl with package availability so the subclasses don't have to implement submit().